onnx/defs/schema.cc

```diff
@@ -724,19 +742,33 @@ OpSchema& OpSchema::FillUsing(const std::function<void(OpSchema&)>& populator) {
   return *this;
 }
 
-void OpSchema::BuildFunction() {
-  function_body_.set_name(this->name_);
-  function_body_.set_doc_string(this->doc_);
-  function_body_.set_since_version(this->since_version_);
-  function_body_.set_status(OperatorStatus(1 - (int)this->support_));
+// only add input to the function body if it is used by a body's node.
+bool FilterFunctionBodyInputByName(const std::string& input_name, const FunctionProto& function_body)
+{
+  for (int i = 0; i< function_body.node_size(); i++)
+  {
+    auto node = function_body.node(i);
+    for (int j = 0; j < node.input_size(); j++)
+      if (node.input(j) == input_name)
+        return true;
+  }
+  return false;
+}
+
+void OpSchema::BuildFunction(FunctionProto& function_body) {
+  function_body.set_name(this->name_);
+  function_body.set_doc_string(this->doc_);
+  function_body.set_since_version(this->since_version_);
+  function_body.set_status(OperatorStatus(1 - (int)this->support_));
   for (auto& i : inputs_) {
-    function_body_.add_input(i.GetName());
+    if (FilterFunctionBodyInputByName(i.GetName(), function_body))
```

Comment:
ONNX compilers and runtimes are required to correctly handle operators with [optional inputs and outputs](https://github.com/onnx/onnx/blob/master/docs/IR.md#optional-inputs-and-outputs). Although functions are somewhat underdocumented, I think it is more consistent that runtimes should be required (and able) to correctly handle function bodies which don't use those optional inputs/outputs. IMO this would be easier to implement too.

Also, this filtering seems like it could be a big source of bugs in the future, and it already has at least two that I can see: it will remove inputs that aren't optional if they aren't used by the function body, and it will remove inputs that are used within a subgraph but not by a node in the top-level function graph.

---

For those reasons, I would strongly recommend this be changed to not filter inputs.

Reply:
totally understand and agree now.